Sources/SourceControl/RepositoryManager.swift

```diff
@@ -165,54 +121,74 @@ public class RepositoryManager {
     /// Get a handle to a repository.
     ///
     /// This will initiate a clone of the repository automatically, if
-    /// necessary, and immediately return. The client can add observers to the
-    /// result in order to know when the repository is available.
-    public func lookup(repository: RepositorySpecifier) -> RepositoryHandle {
-        // Check to see if the repository has been provided.
-        if let handle = repositories[repository.url] {
-            // Fetch and update the repository when it is being looked up.
-            // We need some mechanism to maybe not do this for pinned repos.
+    /// necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping (Result<RepositoryHandle, AnyError>) -> Void) {
+        DispatchQueue.global().async {
+            repostoriesLock.whileLocked {
+                // Check to see if the repository has been provided.
+                if let handle = self.repositories[repository.url] {
+                    // Fetch and update the repository when it is being looked up.
+                    // We need some mechanism to maybe not do this for pinned repos.
+                    do {
+                        let repo = try handle.open()
+                        try repo.fetch()
+                    } catch {
+                        // FIXME: Better error handling.
+                        // Maybe just print a warning that we couldn't fetch it
+                        // since we already have the repo.
+                        print("Error while fetching the repo: \(handle)")
```

Comment:
I'm not sure, it depends on the client... as phrased the API is "give me the up-to-date repo", and we failed at that. I think we could throw an error that communicates this specific information, and clients could downgrade from an error if they want to.